### ISSUE TITLE
ObjectComparator fails on objects containing non empty SplObjectStorage

### ISSUE THREAD

**Issue by nicolopignatelli** [sebastianbergmann/comparator#16]
I'm not sure if this is the intended behavior of the ObjectComparator, but the following code throws a `ComparisonFailure` exception. I suspect it's because of the way SplObjectStorageComparator works.

Can someone confirm this? I'm already investigating in how to fix this, in case the bug is confirmed.

``` php
$storage1 = new \SplObjectStorage();
$storage1->attach(new \stdClass());
$object1 = new SampleClass($storage1, 'b', 'c');

$storage2 = new \SplObjectStorage();
$storage2->attach(new \stdClass());
$object2 = new SampleClass($storage2, 'b', 'c');


$comparator = new ObjectComparator;
$comparator->setFactory(new Factory);

// this throws ComparisonFailure exception
$comparator->assertEquals($object1, $object2);
```

**Comment by nicolopignatelli** [comment]
The SplObjectStorageComparator checks that $actual and $expected contain the same instances using ::contains(). This strategy is enforcing a strict check, which is not how assertEquals() should behave.

This is how PHP 5.5 behaves using only == for comparison:
```php
$class1 = new stdClass();
$storage1 = new SplObjectStorage();
$storage1->attach($class1);
$container1 = new stdClass();
$container1->storage = $storage1;
 
$class2 = new stdClass();
$storage2 = new SplObjectStorage();
$storage2->attach($class2);
$container2 = new stdClass();
$container2->storage = $storage2;
print_r((bool)$container1 == $container2); // echoes 1

**Comment by webmozart** [comment]
I can confirm this. This is due to the usage of `contains()` in SplObjectStorageComparator, which does not allow equal objects (like your `stdClass`) instances but only identical ones.

A solution could be to extend ArrayComparator and to compare the storages like arrays.

**Comment by nicolopignatelli** [comment]
Fixed in #17
